wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql

```diff
@@ -0,0 +1,41 @@
+--
+-- Copyright (c) 2012-2018 Red Hat, Inc.
+-- This program and the accompanying materials are made
+-- available under the terms of the Eclipse Public License 2.0
+-- which is available at https://www.eclipse.org/legal/epl-2.0/
+--
+-- SPDX-License-Identifier: EPL-2.0
+--
+-- Contributors:
+--   Red Hat, Inc. - initial API and implementation
+--
+
+
+-- remove key pair records which linked to non-existing keys
+DELETE FROM che_sign_key_pair kp
```

Comment:
Can't it be simplified like the following?
```
DELETE FROM che_sign_key_pair kp
WHERE NOT EXISTS (
   SELECT id
   FROM   che_sign_key k
   WHERE  k.id = kp.private_key OR k.id = kp.public_key
   )
```

Reply:
Nope.  That query will work only if both private_  and public_ key not exists.

Reply:
I understand. Thanks.

There is a possible case when key par A exists, and it has public key but not private one. Key pair is going to be removed. But private key won't be removed. As result adding foreign key will fail. Am I right? If yes, please consider adding cleaning private and public keys which are referencing non-existing key pair.

Reply:
I understand. Thanks.

There is a possible case when key par A exists, and it has public key but not private one. Key pair is going to be removed. But private key won't be removed. As result adding foreign key will fail. Am I right? If yes, please consider adding cleaning private and public keys which are referencing non-existing key pair.

Reply:
> I don't see reason why foreign key shoud fail?

You're right. I missed that foreign keys are from pair to keys but not wise versa.

Reply:
Added extra unused keys cleanup query

Reply:
https://github.com/eclipse/che/pull/11113/files#diff-f71f338e81773023418fd98cf10db36aR36